funnel/models/membership.py

```diff
@@ -0,0 +1,371 @@
+# -*- coding: utf-8 -*-
+
+from sqlalchemy.ext.associationproxy import association_proxy
+from sqlalchemy.ext.declarative import declared_attr
+from sqlalchemy.ext.hybrid import hybrid_property
+
+from coaster.sqlalchemy import immutable
+
+from . import BaseMixin, UuidMixin, db
+from .profile import Profile
+from .project import Project
+from .user import User
+
+__all__ = ['ProfileAdminMembership', 'ProjectCrewMembership']
+
+
+class ImmutableMembershipMixin(UuidMixin, BaseMixin):
+    """
+    Support class for immutable memberships
+    """
+
+    __uuid_primary_key__ = True
+    #: List of columns that will be copied into a new row when a membership is amended
+    __role_columns__ = ()
+    #: Parent column ('profile_id' or 'project_id' in the subclasses)
+    __parent_column__ = None
+
+    #: Start time of membership, ordinarily a mirror of created_at except
+    #: for records created when the member table was added to the database
+    granted_at = immutable(
+        db.Column(db.TIMESTAMP(timezone=True), nullable=False, default=db.func.utcnow())
+    )
+    #: End time of membership, ordinarily a mirror of updated_at
+    revoked_at = db.Column(db.TIMESTAMP(timezone=True), nullable=True)
+
```

Comment:
Missing column: `is_invite` (default `False`).

Reply:
We forgot about invited memberships entirely in the initial design. Invited memberships are an optional feature that become necessary when an abuse pattern emerges. We have two paths to make invitation-based memberships mandatory:

1. Make invitations a requirement for all members.
2. Make invitations a requirement for some members who have enabled an optional flag in their settings.

The invite flag here solves for path 1 (but not path 2, since it won't stop invite spam). Adding it now makes it easier to turn it on later if required.

Semantics of the flag:

1. When invitations are mandatory, the admin can invite someone to be a member, but can't directly add them. The membership record is created with `is_invite=True, granted_by=current_auth.actor`.

2. The invited user gets a notification (email, push) and a link to an accept or decline page.

3. If the invited user chooses to decline, the existing membership record's `.revoke()` is invoked. This records `revoked_at=db.func.utcnow(), revoked_by=current_auth.actor`. Note that the actor is now the invited user, not the admin who invited them.

4. If the invited user chooses to accept, the existing membership's record's `.replace()` is invoked. This records a revocation as above, plus in the new record `is_invite=False, granted_by=current_auth.actor`. Note that in this case, it will appear that the user granted themselves a membership. The identity of the person who invited them is only available from the previous record. This should be okay as the `granted_by` column is an audit detail and not a product feature.

Reply:
We could do something better with `is_invite` to make it easier to construct a timeline of membership changes. Instead of a boolean, name it `record_type` with one of the following integer values:

1. Add (directly added as a member)
2. Amend (roles were amended)
3. Invite (user was invited)
4. Accept (user accepted an invite)

A decline/revoke type is not necessary as those are captured on the last membership record via the `revoked_at` and `revoked_by` columns.

If we do this, then we can make a timeline entry stating that "User X accepted an invite" instead of "User X granted themselves some roles". However, it will change the filter conditions elsewhere to explicitly exclude item 3, which is awkward. A boolean condition looks cleaner there. A boolean condition could be mimicked by retaining `is_invite` as a hybrid property that has `return self.record_type == MEMBERSHIP_RECORD_TYPE.INVITE`.